README.md
# enum_dispatch

[![crates.io](https://img.shields.io/crates/v/enum_dispatch.svg)](https://crates.io/crates/enum_dispatch)
[![Docs](https://docs.rs/enum_dispatch/badge.svg)](https://docs.rs/enum_dispatch/)
[![License](https://img.shields.io/badge/license-MIT-blue.svg)]()

`enum_dispatch` transforms your trait objects into concrete compound types, increasing their method call speed up to 10x.

## example

If you have the following code...

```rust
// We already have defined MyImplementorA and MyImplementorB, and implemented MyBehavior for each.

trait MyBehavior {
    fn my_trait_method(&self);
}

// Any pointer type -- Box, &, etc.
let a: Box<dyn MyBehavior> = Box::new(MyImplementorA::new());

a.my_trait_method();    //dynamic dispatch
```

...then you can improve its performance using `enum_dispatch` like this:

```rust
#[enum_dispatch]
enum MyBehaviorEnum {
    MyImplementorA,
    MyImplementorB,
}

#[enum_dispatch(MyBehaviorEnum)]
trait MyBehavior {
    fn my_trait_method(&self);
}

let a: MyBehaviorEnum = MyImplementorA::new().into();

a.my_trait_method();    //no dynamic dispatch
```

Notice the differences:

1. The new enum, `MyBehaviorEnum`, whose variants are simply types implementing the trait `MyBehavior`.
2. The new `enum_dispatch` attributes applied to the enum and trait, linking the two together.
3. The removal of the `Box` allocation.
4. Faster trait method calls!

## how to use

0. Add `enum_dispatch` as a Cargo.toml dependency, and `use enum_dispatch::enum_dispatch` in your code.
1. Create a new enum whose variants are any in-scope trait implementors you've defined.
2. Add an `#[enum_dispatch]` attribute to either the enum or trait definition. This will "register" it with the `enum_dispatch` library. Take note of the name of the enum or trait it was applied to -- we'll call it `FirstBlockName`.
3. Add an `#[enum_dispatch(FirstBlockName)]` attribute to the remaining definition. This will "link" it with the previously registered definition.
4. Update your dynamic types to use the new enum instead. You can use `.into()` from any trait implementor to automatically turn it into an enum variant.

## performance

More information on performance can be found in the [docs](https://docs.rs/enum_dispatch/), and benchmarks are available in the `benches` directory.
The following benchmark results give a taste of what can be achieved using `enum_dispatch`.
They compare the speed of repeatedly accessing method calls on a `Vec` of 1024 trait objects of randomized concrete types using either `Box`ed trait objects, `&` referenced trait objects, or `enum_dispatch`ed enum types.

```text
test benches::boxdyn_homogeneous_vec       ... bench:   5,900,191 ns/iter (+/- 95,169)
test benches::refdyn_homogeneous_vec       ... bench:   5,658,461 ns/iter (+/- 137,128)
test benches::enumdispatch_homogeneous_vec ... bench:     479,630 ns/iter (+/- 3,531)
```

## bonus features

### serialization compatibility

While `enum_dispatch` was built with performance in mind, the transformations it applies make all your data structures much more visible to the compiler.
That means you can use [`serde`](https://crates.io/crates/serde) or other similar tools on your trait objects!

### automatic `From` and `TryInto` implementations

`enum_dispatch` will generate a `From` implementation for all inner types to make it easy to instantiate your custom enum.
In addition, it will generate a `TryInto` implementation for all inner types to make it easy to convert back into the original, unwrapped types.

### attribute support

You can use use `#[cfg(...)]` attributes on `enum_dispatch` variants to conditionally include or exclude their corresponding `enum_dispatch` implementations.
Other attributes will be passed straight through to the underlying generated enum, allowing compatibility with other procedural macros.

### `no_std` support

`enum_dispatch` is supported in `no_std` environments.
It's a great fit for embedded devices, where it's super useful to be able to allocate collections of trait objects on the stack.

## tweaks and options

### custom variant names

By default, `enum_dispatch` will expand each enum variant into one with a single unnamed field of the same name as the internal type.
If for some reason you'd like to use a custom name for a particular type in an `enum_dispatch` variant, you can do so as shown below:

```rust
#[enum_dispatch]
enum MyTypes {
    TypeA,
    CustomVariantName(TypeB),
}

let mt: MyTypes = TypeB::new().into();
match mt {
    TypeA(a) => { /* `a` is a TypeA */ },
    CustomVariantName(b) => { /* `b` is a TypeB */ },
}
```

Custom variant names are required for enums and traits with generic type arguments, which can also be optimized by `enum_dispatch`.
Check out [this generics example](tests/generics.rs) to see how that works.

### specify multiple enums at once

If you want to use `enum_dispatch` to implement the same trait for multiple enums, you may specify them all in the same attribute:

```rust
#[enum_dispatch(Widgets, Tools, Gadgets)]
trait CommonFunctionality {
    // ...
}
```

### specify multiple traits at once

Similarly to above, you may use a single attribute to implement multiple traits for a single enum:

```rust
#[enum_dispatch(CommonFunctionality, WidgetFunctionality)]
enum Widget {
    // ...
}
```

### generic enums and traits

`enum_dispatch` can operate on enums and traits with generic parameters.
When linking these, be sure to include the generic parameters in the attribute argument, like below:

```
#[enum_dispatch]
trait Foo<T, U> { /* ... */ }

#[enum_dispatch(Foo<T, U>)]
enum Bar<T: Clone, U: Hash> { /* ... */ }
```

The names of corresponding generic parameters should match between the definition of the enum and trait.

[This example](tests/complex_generics.rs) demonstrates this in more detail.

## troubleshooting

### no impls created?

Be careful not to forget an attribute or mistype the name in a linking attribute.
If parsing is completed before a linking attribute is found, no implementations will be generated.
Due to technical limitations of the macro system, it's impossible to properly warn the user in this scenario.

### can't parse enum?

Types must be fully in scope to be usable as an enum variant. For example, the following will fail to compile:

```rust
#[enum_dispatch]
enum Fails {
    crate::A::TypeA,
    crate::B::TypeB,
}
```

This is because the enum must be correctly parsable before macro expansion. Instead, import the types first:

```rust
use crate::A::TypeA;
use crate::B::TypeB;

#[enum_dispatch]
enum Succeeds {
    TypeA,
    TypeB,
}
```

## technical details

`enum_dispatch` is a procedural macro that implements a trait for a fixed set of types in the form of an enum.
This is faster than using dynamic dispatch because type information will be "built-in" to each enum, avoiding a costly vtable lookup.

Since `enum_dispatch` is a procedural macro, it works by processing and expanding attributed code at compile time.
The folowing sections explain how the example above might be transformed.

### enum processing

There's no way to define an enum whose variants are actual concrete types.
To get around this, `enum_dispatch` rewrites its body by generating a name for each variant and using the provided type as its single tuple-style argument.
The name for each variant isn't particularly important for most purposes, but `enum_dispatch` will currently just use the name of the provided type.

```rust
enum MyBehaviorEnum {
    MyImplementorA(MyImplementorA),
    MyImplementorB(MyImplementorB),
}
```

### trait processing

`enum_dispatch` doesn't actually process annotated traits!
However, it still requires access to the trait definition so it can take note of the trait's name, as well as the function signatures of any methods inside it.

### trait impl creation

Whenever `enum_dispatch` is able to "link" two definitions together, it will generate an `impl` block, implementing the trait for the enum.
In the above example, the linkage is completed by the `MyBehavior` trait definition, so `impl` blocks will be generated directly below that trait.
The generated impl block might look something like this:

```rust
impl MyBehavior for MyBehaviorEnum {
    fn my_trait_method(&self) {
        match self {
            MyImplementorA(inner) => inner.my_trait_method(),
            MyImplementorB(inner) => inner.my_trait_method(),
        }
    }
}
```

Additional trait methods would be expanded accordingly, and additional enum variants would correspond to additional match arms in each method definition.
It's easy to see how quickly this can become unmanageable in manually written code!

### 'From' impl creation

Normally, it would be impossible to initialize one of the new enum variants without knowing its name.
However, with implementations of `From<T>` for each variant, that requirement is alleviated.
The generated implementations could look like the following:

```rust
impl From<MyImplementorA> for MyBehaviorEnum {
    fn from(inner: MyImplementorA) -> MyBehaviorEnum {
        MyBehaviorEnum::MyImplementorA(inner)
    }
}

impl From<MyImplementorB> for MyBehaviorEnum {
    fn from(inner: MyImplementorB) -> MyBehaviorEnum {
        MyBehaviorEnum::MyImplementorB(inner)
    }
}
```

As with above, having a large number of possible type variants would make this very difficult to maintain by hand.

### registry and linkage

Anyone closely familiar with writing macros will know that they must be processed locally, with no context about the surrounding source code.
Additionally, parsed syntax items in `syn` are `!Send` and `!Sync`.
This is for good reason -- with multithreaded compilation and macro expansion, there are no guarantees on the order or lifetime of a reference to any given block of code.
Unfortunately, it also prevents referencing syntax between separate macro invocations.

In the interest of convenience, `enum_dispatch` circumvents these restrictions by converting syntax into a `String` and storing it in `once_cell` lazily initialized `Mutex<HashMap<String, String>>`s whose keys are either the trait or enum names.

There is also a similar `HashMap` dedicated to "deferred" links, since definitions in different files could be encountered in arbitrary orders.
If a linking attribute (with one argument) occurs before the corresponding registry attribute (with no arguments), the argument will be stored as a deferred link.
Once that argument's definition is encountered, impl blocks can be created as normal.

Because of the link deferral mechanism, it's not an error to encounter a linking attribute without being able to implement it.
`enum_dispatch` will simply expect to find the corresponding registry attribute later in parsing.
However, there's no way to insert a callback to check that all deferred links have been processed once all the original source code has been parsed, explaining the impossibility of warning the user of unlinked attributes.


CHANGELOG.md
# Changelog

## 0.3.8

- Preserve attributes from inner fields of enum variants (!27)

## 0.3.7

- Support trait methods with late bound lifetime arguments (#34)

## 0.3.6

- Remove `extra-traits` feature from `syn` dependency (!24)
- Support trait methods with pattern arguments (#44)

## 0.3.5

- Compatibility with `syn >= 1.0.58` (#37, !21)

## 0.3.4

- Support enum variants named `Error` (#36)

## 0.3.3

### Compatibility warning
Users who had previously used an `#[enum_dispatch(...)]` attribute containing the name of a generic trait will need to update the attribute to include matching generic arguments.
See #35 for details.

- Support trait methods with generic type parameters (#28)
- Officially support linking generic traits (#26)

## 0.3.2

- Support `cfg` attributes on enum variants and trait items (#24)

## 0.3.1

- Support multiple comma separated traits or enums in `enum_dispatch` attributes (#3, !14)
- Pass attributes from trait methods to the generated implementations (!15)

## 0.3.0

Rerelease of 0.2.3 to undo unintentional semver incompatibility (#16)

## 0.2.4

Rerelease of 0.2.2 to undo unintentional semver incompatibility (#16)

## 0.2.3

- Support identical method names across traits and base structs (!12)

## 0.2.2

- Support multiple trait implementations per enum (!13)

## 0.2.1

- Pass attributes from enum variants to the generated enums (#14)
- Support enum variants with generic parameters (!4)

## 0.2.0

- Generate implementations of `TryInto` instead of `TryFrom`, which cannot be implemented on foreign types (#10)


tests/no_std.rs
#![no_std]

use enum_dispatch::enum_dispatch;
use core::convert::TryInto;

struct MyImplementorA {
}

impl MyBehavior for MyImplementorA {
    fn my_trait_method(&self) {
    }
}

struct MyImplementorB {
}

impl MyBehavior for MyImplementorB {
    fn my_trait_method(&self) {
    }
}

#[enum_dispatch]
enum MyBehaviorEnum {
    MyImplementorA,
    MyImplementorB,
}

#[enum_dispatch(MyBehaviorEnum)]
trait MyBehavior {
    fn my_trait_method(&self);
}

#[test]
fn main() {
    let a: MyBehaviorEnum = MyImplementorA {}.into();
    a.my_trait_method();

    let _a: MyImplementorA = a.try_into().unwrap();
}


tests/generics.rs
use enum_dispatch::enum_dispatch;

pub struct SuperFoo<T: Bar> {
    _bar: T,
}

impl<T: Bar> Foo<T> for SuperFoo<T> {
    fn do_something(&mut self, _val: T) {
        println!("SuperFoo");
    }
}

pub struct UltraFoo<T: Bar> {
    _bar: T,
}

impl<T: Bar> Foo<T> for UltraFoo<T> {
    fn do_something(&mut self, _val: T) {
        println!("UltraFoo");
    }
}

pub trait Bar {}

#[enum_dispatch]
pub trait Foo<T: Bar> {
    fn do_something(&mut self, val: T);
}

#[enum_dispatch(Foo)]
pub enum AnyFoo<T: Bar> {
    SuperFoo(SuperFoo<T>),
    UltraFoo(UltraFoo<T>),
}

#[enum_dispatch]
pub trait Faz {}

use std::marker::PhantomData;
pub struct SuperFaz<'a>(PhantomData<&'a u8>);
pub struct UltraFaz();

impl<'a> Faz for SuperFaz<'a> {}
impl Faz for UltraFaz {}

#[enum_dispatch(Faz)]
pub enum AnyFaz<'a> {
    SuperFaz(SuperFaz<'a>),
    UltraFaz(UltraFaz),
}


src/attributed_parser.rs
//! Contains helper utilities for parsing items that have been annotated with the `enum_dispatch`
//! procedural macro attribute.
use crate::enum_dispatch_item;
use crate::proc_macro;

/// Enumerates all successful results of parsing an `enum_dispatch` annotated syntax block.
#[derive(Clone)]
pub enum ParsedItem {
    Trait(syn::ItemTrait),
    EnumDispatch(enum_dispatch_item::EnumDispatchItem),
}

/// Parses any syntax item that was annotated with the `enum_dispatch` attribute and returns its
/// itemized results.
pub fn parse_attributed(item: proc_macro::TokenStream) -> Result<ParsedItem, ()> {
    if let Ok(enumdef) = syn::parse(item.clone()) {
        Ok(ParsedItem::EnumDispatch(enumdef))
    } else if let Ok(traitdef) = syn::parse(item) {
        Ok(ParsedItem::Trait(traitdef))
    } else {
        Err(())
    }
}


src/lib.rs
//! `enum_dispatch` provides a set of macros that can be used to easily refactor dynamically
//! dispatched trait accesses to improve their performance by up to 10x.
//!
//! Accessing structures through dynamic dispatch is known to have a high runtime cost. Dynamic
//! dispatch is traditionally used to hide unnecessary type information, improving encapsulation
//! and making it trivial to add new implementations. However, this hiding of information means
//! that each time a structure is dynamically accessed, the program must perform a lookup of the
//! type's information in a virtual table. The extra round-trips to the vtable quickly add up.
//!
//! In Rust, dynamic dispatch is done using traits. Rust 2018 adds the `impl` and `dyn` keywords to
//! make it easier to keep track of instances of dynamic dispatch, but it's not always easy to
//! avoid it entirely.
//!
//! # Feature documentation
//!
//! For full documentation of features like generic support, custom variant names, and more, please
//! check the repository's
//! [README](https://gitlab.com/antonok/enum_dispatch/-/blob/master/README.md).
//!
//! # How it works
//!
//! Observe the following example of code describing a user interface with knobs. Each knob can
//! hold a value between 0.0 and 1.0. Some knobs provide a *linear* range, whereas other knobs
//! provide a *logarithmic* range.
//!
//! ```
//! trait KnobControl {
//!     fn set_position(&mut self, value: f64);
//!     fn get_value(&self) -> f64;
//! }
//!
//! struct LinearKnob {
//!     position: f64,
//! }
//!
//! struct LogarithmicKnob {
//!     position: f64,
//! }
//!
//! impl KnobControl for LinearKnob {
//!     fn set_position(&mut self, value: f64) {
//!         self.position = value;
//!     }
//!
//!     fn get_value(&self) -> f64 {
//!         self.position
//!     }
//! }
//!
//! impl KnobControl for LogarithmicKnob {
//!     fn set_position(&mut self, value: f64) {
//!         self.position = value;
//!     }
//!
//!     fn get_value(&self) -> f64 {
//!         (self.position + 1.).log2()
//!     }
//! }
//!
//! fn main() {
//!     let v: Vec<Box<dyn KnobControl>> = vec![
//!         //set the knobs
//!     ];
//!
//!     //use the knobs
//! }
//! ```
//!
//! There are other ways to keep an arbitrarily ordered list of different knob types, but none of
//! them are quite as simple or easy to maintain. Unfortunately, this implementation uses both heap
//! allocated `Box`es and dynamic dispatch, which will have performance implications.
//!
//! One alternative is to use introduce a new enum type that can hold either a `LinearKnob` or a
//! `LogarithmicKnob` as a variant, and also implements `KnobControl` by matching on itself and
//! delegating calls to its variants. This would look like the following:
//!
//! ```
//! # trait KnobControl {
//! #     fn set_position(&mut self, value: f64);
//! #     fn get_value(&self) -> f64;
//! # }
//! #
//! # struct LinearKnob {
//! #     position: f64,
//! # }
//! #
//! # struct LogarithmicKnob {
//! #     position: f64,
//! # }
//! #
//! # impl KnobControl for LinearKnob {
//! #     fn set_position(&mut self, value: f64) {
//! #         self.position = value;
//! #     }
//! #
//! #     fn get_value(&self) -> f64 {
//! #         self.position
//! #     }
//! # }
//! #
//! # impl KnobControl for LogarithmicKnob {
//! #     fn set_position(&mut self, value: f64) {
//! #         self.position = value;
//! #     }
//! #
//! #     fn get_value(&self) -> f64 {
//! #         (self.position + 1.).log2()
//! #     }
//! # }
//! #
//! enum Knob {
//!     Linear(LinearKnob),
//!     Logarithmic(LogarithmicKnob),
//! }
//!
//! impl KnobControl for Knob {
//!     fn set_position(&mut self, value: f64) {
//!         match self {
//!             Knob::Linear(inner_knob) => inner_knob.set_position(value),
//!             Knob::Logarithmic(inner_knob) => inner_knob.set_position(value),
//!         }
//!     }
//!
//!     fn get_value(&self) -> f64 {
//!         match self {
//!             Knob::Linear(inner_knob) => inner_knob.get_value(),
//!             Knob::Logarithmic(inner_knob) => inner_knob.get_value(),
//!         }
//!     }
//! }
//! ```
//!
//! Performance with this implementation is significantly improved, since all the information the
//! program could possibly need to know about each knob can be deduced at compile time. Besides
//! avoiding heap allocations and vtable lookups, this allows the compiler to squeeze out even more
//! optimization through function inlining.
//!
//! However, it's easy to see that the cost of maintaining the source code for this extra structure
//! is quite high. What happens when we add more knob types? What happens when we add more trait
//! methods? Even worse, what happens when we do both!
//!
//! The resulting code is very repetitive, but that makes it a great target for automatic
//! generation. The `enum_dispatch` macro can do the automatic generation for you. Examine the code
//! to generate the same implementation when using `enum_dispatch`.
//!
//! ```
//! # use enum_dispatch::enum_dispatch;
//! #
//! #[enum_dispatch]
//! trait KnobControl {
//!     //...
//! #     fn set_position(&mut self, value: f64);
//! #     fn get_value(&self) -> f64;
//! }
//! #
//! # struct LinearKnob {
//! #     position: f64,
//! # }
//! #
//! # struct LogarithmicKnob {
//! #     position: f64,
//! # }
//! #
//! # impl KnobControl for LinearKnob {
//! #     fn set_position(&mut self, value: f64) {
//! #         self.position = value;
//! #     }
//! #
//! #     fn get_value(&self) -> f64 {
//! #         self.position
//! #     }
//! # }
//! #
//! # impl KnobControl for LogarithmicKnob {
//! #     fn set_position(&mut self, value: f64) {
//! #         self.position = value;
//! #     }
//! #
//! #     fn get_value(&self) -> f64 {
//! #         (self.position + 1.).log2()
//! #     }
//! # }
//!
//! #[enum_dispatch(KnobControl)]
//! enum Knob {
//!     LinearKnob,
//!     LogarithmicKnob,
//! }
//! ```
//!
//! That's it. `enum_dispatch` will also automatically generate implementations of
//! `std::convert::From` for each enum variant, so that new `Knob`s can be created without concern
//! for the names of each enum variant.
//!
//! ```
//! # use enum_dispatch::enum_dispatch;
//! #
//! # #[enum_dispatch]
//! # trait KnobControl {
//! #     fn set_position(&mut self, value: f64);
//! #     fn get_value(&self) -> f64;
//! # }
//! #
//! # struct LinearKnob {
//! #     position: f64,
//! # }
//! #
//! # struct LogarithmicKnob {
//! #     position: f64,
//! # }
//! #
//! # impl KnobControl for LinearKnob {
//! #     fn set_position(&mut self, value: f64) {
//! #         self.position = value;
//! #     }
//! #
//! #     fn get_value(&self) -> f64 {
//! #         self.position
//! #     }
//! # }
//! #
//! # impl KnobControl for LogarithmicKnob {
//! #     fn set_position(&mut self, value: f64) {
//! #         self.position = value;
//! #     }
//! #
//! #     fn get_value(&self) -> f64 {
//! #         (self.position + 1.).log2()
//! #     }
//! # }
//! #
//! # #[enum_dispatch(KnobControl)]
//! # enum Knob {
//! #     LinearKnob,
//! #     LogarithmicKnob,
//! # }
//! #
//! # fn some_existing_knobs() -> (LinearKnob, LogarithmicKnob) {
//! #     (LinearKnob { position: 0.5 }, LogarithmicKnob { position: 0.5 })
//! # }
//! #
//! let (a_linear_knob, a_logarithmic_knob) = some_existing_knobs();
//!
//! let knob = Knob::from(a_linear_knob);
//! let knob = Knob::from(a_logarithmic_knob);
//! ```
//!
//! # Performance
//!
//! The `benches` directory contains three benchmarks of different natures, each comparing four
//! different methods of accessing a traited struct of an arbitrary type. The four methods are as
//! follows:
//!
//! | test name    | explanation                                                                             |
//! |--------------|-----------------------------------------------------------------------------------------|
//! | boxdyn       | The easiest way to access a struct, using a heap allocation and dynamic dispatch.       |
//! | refdyn       | Accesses the struct by reference, but still using dynamic dispatch. No heap allocation. |
//! | customderive | Uses a similar macro approach from the external [`enum_derive`](https://github.com/DanielKeep/rust-custom-derive) crate, which implements a method that returns an inner type as a dynamic trait object. |
//! | enumdispatch | Implemented using this crate.                                                           |
//!
//! ## The benchmarks
//!
//! The following benchmark results were measured on a Ryzen 7 2700x CPU.
//!
//! ### compiler_optimized
//!
//! The first set of benchmarks creates trait objects and measures the speed of accessing a method
//! on them.
//!
//! ```text
//! test benches::boxdyn_compiler_optimized       ... bench:   2,135,418 ns/iter (+/- 12,575)
//! test benches::customderive_compiler_optimized ... bench:   2,611,860 ns/iter (+/- 18,644)
//! test benches::enumdispatch_compiler_optimized ... bench:           0 ns/iter (+/- 0)
//! test benches::refdyn_compiler_optimized       ... bench:   2,132,591 ns/iter (+/- 22,114)
//! ```
//!
//! It's easy to see that `enum_dispatch` is the clear winner here!
//!
//! Ok, fine. This wasn't a fair test. The compiler is able to "look through" the trait method call
//! in the enum_dispatch case, notices that the result is unused, and removes it as an
//! optimization. However, this still highlights an important property of `enum_dispatch`ed types:
//! the compiler is able to infer much better optimizations when possible.
//!
//! ### blackbox
//!
//! The next set of benchmarks uses the `test::black_box` method to hide the fact that the result
//! of the method is unused.
//!
//! ```text
//! test benches::boxdyn_blackbox       ... bench:   2,131,736 ns/iter (+/- 24,937)
//! test benches::customderive_blackbox ... bench:   2,611,721 ns/iter (+/- 23,502)
//! test benches::enumdispatch_blackbox ... bench:     471,740 ns/iter (+/- 1,439)
//! test benches::refdyn_blackbox       ... bench:   2,131,978 ns/iter (+/- 21,547)
//! ```
//!
//! The competitors faced virtually no impact, whereas `enum_dispatch` takes the full force of the
//! `black_box` call. This test shows the power that avoiding dynamic dispatch gives to the
//! compiler in the context of the previous test, but also demonstrates how much faster
//! `enum_dispatch` is in real code: almost 5 times faster than the closest alternative.
//!
//! ### homogenous_vec
//!
//! The final set of benchmarks puts 1024 traited structs of arbitrary types at random into a `Vec`
//! and measures the time it takes to successively iterate over the entire `Vec`, calling
//! `black_box`ed methods on each element.
//!
//! ```text
//! test benches::boxdyn_homogeneous_vec       ... bench:   5,900,191 ns/iter (+/- 95,169)
//! test benches::customderive_homogeneous_vec ... bench:   4,831,808 ns/iter (+/- 140,437)
//! test benches::enumdispatch_homogeneous_vec ... bench:     479,630 ns/iter (+/- 3,531)
//! test benches::refdyn_homogeneous_vec       ... bench:   5,658,461 ns/iter (+/- 137,128)
//! ```
//!
//! This might be one of the most likely use cases for traited structs of arbitrary types, and it's
//! where `enum_dispatch` really shines. Since a `Vec` of `enum_dispatch` objects is actually a
//! `Vec` of enums rather than addresses, accessing an element takes half the indirection of the
//! other techniques. Add that to the lack of vtable accesses, and we have a result that is 10
//! times faster than the closest alternative, and almost 12 times faster than the best technique
//! from the standard library.

#![doc(
    html_logo_url = "https://gitlab.com/antonok/enum_dispatch/raw/master/enum_dispatch.svg",
    html_favicon_url = "https://gitlab.com/antonok/enum_dispatch/raw/master/enum_dispatch.svg"
)]

extern crate proc_macro;

use proc_macro2::TokenStream;
use quote::{ToTokens, TokenStreamExt};

/// Used for converting a macro input into an ItemTrait or an EnumDispatchItem.
mod attributed_parser;
/// Provides local storage for enum and trait definitions so that they can be accessed later.
mod cache;
/// Provides a custom syntax specification for the arguments to an `#[enum_dispatch(...)]` attribute.
mod enum_dispatch_arg_list;
/// Provides a custom syntax specification for enum dispatch syntax blocks.
mod enum_dispatch_item;
/// Provides a custom syntax specification for the variants of enum dispatch syntax blocks.
mod enum_dispatch_variant;
/// Provides utilities for building enum dispatch implementations.
mod expansion;
/// Convenience trait for token parsing.
mod filter_attrs;
/// Convenience methods for constructing `syn` types.
mod syn_utils;

use crate::expansion::add_enum_impls;

/// Annotating a trait or enum definition with an `#[enum_dispatch]` attribute will register it
/// with the enum_dispatch library, allowing it to be used to generate impl blocks elsewhere.
///
/// Annotating a trait or enum definition with an `#[enum_dispatch(BlockName)]` attribute will
/// generate an enum dispatch implementation for the specified trait/enum pair, as well as adding
/// an impl of std::convert::From for each variant. When annotating a trait, BlockName should be the
/// name of a registered enum. When annotating an enum, BlockName should be the name of a registered
/// trait.
///
/// An annotated enum should have variants that are simply the names of types imported to the
/// current scope. To force individual variants to use a custom name when expanded, each variant
/// can also take the form of a normal tuple-style enum variant with a single field.
#[proc_macro_attribute]
pub fn enum_dispatch(attr: proc_macro::TokenStream, item: proc_macro::TokenStream) -> proc_macro::TokenStream {
    enum_dispatch2(attr.into(), item.into()).into()
}

/// `proc_macro2::TokenStream` compatible version of the `enum_dispatch` function.
///
/// Using only `proc_macro2::TokenStream` inside the entire crate makes methods unit-testable and
/// removes the need for conversions everywhere.
fn enum_dispatch2(attr: TokenStream, item: TokenStream) -> TokenStream {
    let new_block = attributed_parser::parse_attributed(item.clone()).unwrap();
    let mut expanded = match &new_block {
        attributed_parser::ParsedItem::Trait(traitdef) => {
            cache::cache_trait(traitdef.to_owned());
            item
        }
        attributed_parser::ParsedItem::EnumDispatch(enumdef) => {
            cache::cache_enum_dispatch(enumdef.clone());
            syn::ItemEnum::from(enumdef.to_owned())
                .into_token_stream()
        }
    };
    // If the attributes are non-empty, the new block should be "linked" to the listed definitions.
    // Those definitions may or may not have been cached yet.
    // If one is not cached yet, the link will be pushed into the cache, and impl generation will
    // be deferred until the missing definition is encountered.
    // For now, we assume it is already cached.
    if !attr.is_empty() {
        syn::parse2::<enum_dispatch_arg_list::EnumDispatchArgList>(attr)
            .expect("Could not parse arguments to `#[enum_dispatch(...)]`.")
            .arg_list
            .into_iter()
            .for_each(|p| {
                if p.leading_colon.is_some() || p.segments.len() != 1 {
                    panic!("Paths in `#[enum_dispatch(...)]` are not supported.");
                }
                let syn::PathSegment {
                    ident: attr_name,
                    arguments: attr_generics
                } = p.segments.last().unwrap();
                let attr_generics = match attr_generics.clone() {
                    syn::PathArguments::None => vec![],
                    syn::PathArguments::AngleBracketed(args) => {
                        assert!(args.colon2_token.is_none());
                        args.args.iter().map(|generic_arg| {
                            match generic_arg {
                                syn::GenericArgument::Type(syn::Type::Path(t)) if t.qself.is_none() => Some(t.path.get_ident().expect("Generic binding paths in #[enum_dispatch(...)] are not supported").clone()),
                                syn::GenericArgument::Type(syn::Type::Infer(_)) => None,
                                syn::GenericArgument::Type(_) => panic!("Generics in #[enum_dispatch(...)] must be identifiers"),
                                syn::GenericArgument::Lifetime(_) => panic!("Lifetime generics in #[enum_dispatch(...)] are not supported"),
                                syn::GenericArgument::Binding(_) => panic!("Generic equality constraints in #[enum_dispatch(...)] are not supported"),
                                syn::GenericArgument::Constraint(_) => panic!("Generic trait constraints in #[enum_dispatch(...)] are not supported"),
                                syn::GenericArgument::Const(_) => panic!("Const expression generics in #[enum_dispatch(...)] are not supported"),
                            }
                        }).collect::<Vec<_>>()
                    }
                    syn::PathArguments::Parenthesized(_) => panic!("Expected angle bracketed generic arguments, found parenthesized arguments"),
                };
                match &new_block {
                    attributed_parser::ParsedItem::Trait(traitdef) => {
                        cache::defer_link((attr_name, attr_generics.len()), (&traitdef.ident, traitdef.generics.type_params().count()))
                    }
                    attributed_parser::ParsedItem::EnumDispatch(enumdef) => {
                        cache::defer_link((attr_name, attr_generics.len()), (&enumdef.ident, enumdef.generics.type_params().count()))
                    }
                }
            });
    };
    // It would be much simpler to just always retrieve all definitions from the cache. However,
    // span information is not stored in the cache. Saving the newly retrieved definition prevents
    // *all* of the span information from being lost.
    match new_block {
        attributed_parser::ParsedItem::Trait(traitdef) => {
            let additional_enums =
                cache::fulfilled_by_trait(&traitdef.ident, traitdef.generics.type_params().count());
            for enumdef in additional_enums {
                expanded.append_all(add_enum_impls(enumdef, traitdef.clone()));
            }
        }
        attributed_parser::ParsedItem::EnumDispatch(enumdef) => {
            let additional_traits =
                cache::fulfilled_by_enum(&enumdef.ident, enumdef.generics.type_params().count());
            for traitdef in additional_traits {
                expanded.append_all(add_enum_impls(enumdef.clone(), traitdef));
            }
        }
    }
    expanded
}
